packages/vite/src/node/plugins/importAnalysisBuild.ts

```diff
@@ -74,11 +74,12 @@ function preload(
       // @ts-ignore
       seen[dep] = true
       const isCss = dep.endsWith('.css')
-      const cssSelector = isCss ? '[rel="stylesheet"]' : ''
-      // @ts-ignore check if the file is already preloaded by SSR markup
-      if (document.querySelector(`link[href="${dep}"]${cssSelector}`)) {
-        return
-      }
+      const allLinks = Array.from(document.querySelectorAll('link'))
+      const isPreloaded = allLinks.some((link) => {
+        return link.href === dep && (!isCss || link.rel === 'stylesheet')
+      })
+      if (isPreloaded) return
```

Comment:
Even with the other changes, I think we are missing something here. The bug you reported was because the CSS was added in the HTML through a different path so you should still be able to see the issue IIUC.
I think we need something similar to what @Tal500 did here https://github.com/vitejs/vite/pull/9920/files#diff-121079017d1fa98d6a0ea3354d5c73df45ab277078228cd7384c7a4e84c5a813R88 (as they commented in your issue).
But that code may also have a bug because the `href` would be relative to the HTML and not to the `importerUrl`. I think we need to compare (only when using relative base) the `dep` to `new URL(link.href,location.href).href` 🤔

Reply:
> Even with the other changes, I think we are missing something here. The bug you reported was because the CSS was added in the HTML through a different path so you should still be able to see the issue IIUC. I think we need something similar to what @Tal500 did here https://github.com/vitejs/vite/pull/9920/files#diff-121079017d1fa98d6a0ea3354d5c73df45ab277078228cd7384c7a4e84c5a813R88 (as they commented in your issue). But that code may also have a bug because the `href` would be relative to the HTML and not to the `importerUrl`. I think we need to compare (only when using relative base) the `dep` to `new URL(link.href,location.href).href` 🤔

Here's some confession - I wanted to do full absolute path resolving via `new URL(link.href,location.href).href` in both side of the equations - but when I did so, some of the automatic tests were failing (not related to legacy)...
So I didn't want to have a bad impression, therefore I slipped it under the hood, and told myself I'll deal with this issue later, or explain it if I'd be asked (tried to keep things simple).
See the commitment of shame here - [e72c0905bec8b0540fa41622956d26028beed04e](https://github.com/vitejs/vite/pull/9920/commits/e72c0905bec8b0540fa41622956d26028beed04e). In the commitment before it, it was implemented both my ideas and yours, comparing absolute path from both sides of the equations.
Need probably to fix the automatic tests, see the failing tests [here](https://github.com/vitejs/vite/actions/runs/2988945070).